src/chronify/column_representative_time_range_generator.py

```diff
@@ -0,0 +1,116 @@
+
+
+class ColumnRepresentativeHandlerBase:
+    """
+    Base class for handling timestamps in different Column Representative formats.
+
+    Methods
+    -------
+    list_distinct_timestamps_from_dataframe
+        list the unique timestamps as a tuple of integers (for example (2024, 12, 31, 7) for year, month, day, hour)
+        given a dataframe of time representing columns
+    """
+
+    @abc.abstractmethod
+    def _iter_timestamps(self) -> Generator[tuple[int, ...], None, None]:
+        """Iterates over a tuples that represent times from column schema."""
+
+    @abc.abstractmethod
+    def list_distinct_timestamps_from_dataframe(self, df: pd.DataFrame) -> list[tuple[int, ...]]:
+        """Returns all unique tuples of representative column time, sorted by longest to shortest duration columns (year > month > day > hour)."""
+
+
+class ColumnRepresentativeHandlerHourly(ColumnRepresentativeHandlerBase):
+    def __init__(self, model: YearMonthDayHourTimeNTZ | MonthDayHourTimeNTZ, year: int) -> None:
+        self._model = model
+        self._year = year
+        self._time_columns = self._model.list_time_columns()
+
+    def _iter_timestamps(self) -> Generator[tuple[int, ...], None, None]:
+        for dt in pd.date_range(
+            start=datetime(self._year, 1, 1),
+            periods=self._model.unique_timestamps_length,
+            freq="1h",
+        ):
+            if isinstance(self._model, YearMonthDayHourTimeNTZ):
+                yield dt.year, dt.month, dt.day, dt.hour + 1
+            elif isinstance(self._model, MonthDayHourTimeNTZ):
+                yield dt.month, dt.day, dt.hour + 1
+
+    def list_distinct_timestamps_from_dataframe(self, df: pd.DataFrame) -> list[tuple[int, ...]]:
+        df = df[self._time_columns].astype(int).drop_duplicates().sort_values(self._time_columns)
+        return df.to_records(index=False).tolist()  # type: ignore
+
+
+class ColumnRepresentativeHandlerPeriod(ColumnRepresentativeHandlerBase):
+    def __init__(self, model: YearMonthDayPeriodTimeNTZ, year: int) -> None:
+        self._model = model
+        self._year = year
+        self._time_columns = self._model.list_time_columns()
+
+    def _iter_timestamps(self) -> Generator[tuple[int, ...], None, None]:
+        for dt in pd.date_range(
+            start=datetime(self._year, 1, 1),
+            periods=self._model.unique_timestamps_length,
+            freq="1D",
+        ):
+            yield dt.year, dt.month, dt.day
+
+    def list_distinct_timestamps_from_dataframe(self, df: pd.DataFrame) -> list[tuple[int, ...]]:
+        int_columns = [
+            self._model.year_column,
+            self._model.month_column,
+            self._model.day_column,
+        ]
+        assert all(
+            col in df.columns for col in int_columns
+        ), f"Not all {int_columns=}, are in the input dataframe {df.columns=}"
+        df = df[int_columns].astype(int).drop_duplicates().sort_values(int_columns)
```

Comment:
What is the reason that casting to int is necessary? Maybe I missed a response...the last I recall, it was an unknown reason.